docs/source/developer_guide/testing_framework.rst

```diff
@@ -0,0 +1,269 @@
+.. _testing_framework:
+
+``Sktime`` testing framework overview
+=====================================
+
+``sktime`` uses ``pytest`` for testing interface compliance of estimators, and correctness of code.
+This page gives an overview over the tests, and introductions on how to add tests, or how to extend the testing framework.
+
+.. contents::
+   :local:
+
+Test module architecture
+------------------------
+
+``sktime`` testing happens on three layers, roughly corresponding to the inheritance layers of estimators.
+
+* "package level": testing interface compliance with the ``BaseObject`` and ``BaseEstimator`` specifications, in
+``tests/test_all_estimators.py``
+* "module level": testing interface compliance of concrete estimators with their scitype base class, for instance
+``forecasting/tests/test_all_forecasters.py``
+* "low level": testing individual functionality of estimators or other code, in individual files in ``tests`` folders
+
+Module conventions are as follows:
+
+* each module contains a ``tests`` folder, which contains tests specific to that module.
+Sub-modules may also contain ``tests`` folders.
+* ``tests`` folders may contain ``_config.py`` files to collect test configuration settings for that module
+* generic utilities for tests are located in the module ``utils._testing``.
+ Tests for these utilities should be contained in the ``utils._testing.tests`` folder.
+* each test module corresponding to a learning task and estimator scitype should contain
+module level tests in a test ``test_all_[name_of_scitype].py`` file that tests interface compliance of all estimators adhering to the scitype.
+ For instance, ``forecasting/tests/test_all_forecasters.py``, or ``distances/tests/test_all_dist_kernels.py``.
+* Learning task specific tests should not duplicate package level, generic estimator tests in ``test_all_estimators.py``
+
+Test code architecture
+----------------------
+
+.. _pytestuse: https://docs.pytest.org/en/6.2.x/example/index.html
+
+``sktime`` test files should use best ``pytest`` practice such as fixtures or test parameterization where possible,
+instead of custom logic, see `pytest documentation <pytestuse>`_.
+
+Estimator tests use ``sktime``'s framework plug-in to ``pytest_generate_tests``,
+which parameterizes estimator fixtures and data input scenarios.
+
+An illustrative example
+~~~~~~~~~~~~~~~~~~~~~~~
+
+Starting with an example:
+
+.. code-block::
+
+   def test_fit_returns_self(estimator_instance, scenario):
+      """Check that fit returns self."""
+      fit_return = scenario.run(estimator_instance, method_sequence=["fit"])
+      assert (
+         fit_return is estimator_instance
+      ), f"Estimator: {estimator_instance} does not return self when calling fit"
+
+Here, ``estimator_instance`` and ``scenario`` fixture are looped over, assuming
```

Comment:
Where exactly is the loop?

Reply:
the loop is carried out by `pytest`, via fixture paramterization (see there)

Reply:
added more explanation here